torch_geometric/nn/models/GraphMask.py

```diff
@@ -0,0 +1,839 @@
+                output = self.full_biases[i]
+                for j in range(len(gate_input)):
+                    partial = self.gates[i * 4][j](gate_input[j][i])
+                    result = self.gates[(i * 4) + 1][j](partial)
+                    output = output + result
+                relu_output = self.gates[(i * 4) + 2](output / len(gate_input))
+                sampling_weights = self.gates[(i * 4) +
+                                              3](relu_output).squeeze(dim=-1)
+                sampling_weights, _ = self.__hard_concrete__(
+                    sampling_weights, training=False)
+                if i == 0:
+                    edge_weight = sampling_weights
+                else:
+                    if (edge_weight.size(-1) != sampling_weights.size(-1)
+                            and self.layer_type == 'GAT'):
+                        sampling_weights = F.pad(
+                            input=sampling_weights,
+                            pad=(0, edge_weight.size(-1) -
+                                 sampling_weights.size(-1), 0, 0),
+                            mode='constant', value=0)
+                    edge_weight = torch.cat((edge_weight, sampling_weights), 0)
+                if self.log:  # pragma: no cover
+                    pbar.update(1)
+        if self.log:  # pragma: no cover
+            pbar.close()
+
+        edge_mask = edge_weight.view(-1,
+                                     edge_weight.size(0) // self.num_layers)
+        edge_mask = torch.mean(edge_mask, 0)
+
+        return edge_mask
+
+    def __train_graph_explainer__(self, graph_idx, x, edge_index, **kwargs):
+        r"""Learns a node feature mask and an edge mask and returns only the
+        learned node feature mask that plays a crucial role to explain the
+        prediction made by the GNN for graph :attr:`graph_idx`.
+
+        Args:
+            graph_idx (int): Graph to explain.
+            x (Tensor): The node feature matrix for graph :attr:`graph_idx`.
+            edge_index (LongTensor): The edge indices for graph
+                :attr:`graph_idx`.
+            **kwargs (optional): Additional arguments passed to the GNN module.
+
+        :rtype: (:class:`Tensor`)
+        """
+
+        if not isinstance(graph_idx, int):
+            raise ValueError(
+                "'graph_idx' parameter must be of type 'integer'.")
+
+        self.freeze_model(self.model)
+        self.__set_flags__(self.model)
+        batch = torch.zeros(x.shape[0], dtype=int, device=x.device)
+
+        input_dims, output_dims = [], []
+        for module in self.model.modules():
+            if isinstance(module, MessagePassing):
+                input_dims.append(module.in_channels)
+                output_dims.append(module.out_channels)
+
+        self.__set_masks__(input_dims, output_dims, output_dims, x)
+
+        optimizer = torch.optim.Adam(self.parameters(), lr=self.lr)
```

Comment:
Actually, I've tried many different optimizers to see which one fits the best to this explainer and found ADAM worked the best. So, have put this optimizer inside the explainer function here. Even this is what had been brought to practice for GNNExplainer implementation of PyG [here](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/models/gnn_explainer.py) if you check inside `explain_graph` function on Line 157. Moreover, I personally don't support the notion of providing user with the option to choose any optimizer they want their explainer to optimize with as this will unnecessarily increase the number of parameters for the explainer function. Hope, this makes some sense now.